pkg/secret/secret_monitor.go

```diff
@@ -0,0 +1,217 @@
+}
+
+type monitoredItem struct {
+	itemMonitor *singleItemMonitor
+	numHandlers atomic.Int32
+}
+
+// secretMonitor is an implementation of the SecretMonitor
+type secretMonitor struct {
+	kubeClient kubernetes.Interface
+	lock       sync.RWMutex
+	monitors   map[ObjectKey]*monitoredItem
+}
+
+func NewSecretMonitor(kubeClient kubernetes.Interface) SecretMonitor {
+	return &secretMonitor{
+		kubeClient: kubeClient,
+		monitors:   map[ObjectKey]*monitoredItem{},
+	}
+}
+
+// AddSecretEventHandler adds a secret event handler to the monitor.
+func (s *secretMonitor) AddSecretEventHandler(ctx context.Context, namespace, secretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
+	return s.addSecretEventHandler(ctx, namespace, secretName, handler, s.createSecretInformer(namespace, secretName))
+}
+
+// createSecretInformer creates a SharedInformer for monitoring a specific secret.
+func (s *secretMonitor) createSecretInformer(namespace, name string) func() cache.SharedInformer {
+	return func() cache.SharedInformer {
+		return cache.NewSharedInformer(
+			cache.NewListWatchFromClient(
+				s.kubeClient.CoreV1().RESTClient(),
+				"secrets",
+				namespace,
+				fields.OneTermEqualSelector("metadata.name", name),
+			),
+			&corev1.Secret{},
+			0,
+		)
+	}
+}
+
+// addSecretEventHandler adds a secret event handler and starts the informer if not already running.
+func (s *secretMonitor) addSecretEventHandler(ctx context.Context, namespace, secretName string, handler cache.ResourceEventHandler, createInformerFn func() cache.SharedInformer) (SecretEventHandlerRegistration, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if handler == nil {
+		return nil, fmt.Errorf("nil handler is provided")
+	}
+
+	// secret identifier (namespace/secret)
+	key := NewObjectKey(namespace, secretName)
+
+	// start secret informer if monitor does not exists
+	m, exists := s.monitors[key]
+	if !exists {
+		m = &monitoredItem{}
+		sharedInformer := createInformerFn()
+		m.itemMonitor = newSingleItemMonitor(key, sharedInformer)
+		go m.itemMonitor.StartInformer(ctx)
+
+		// wait for first sync
+		if !cache.WaitForCacheSync(context.Background().Done(), m.itemMonitor.HasSynced) {
+			return nil, fmt.Errorf("failed waiting for cache sync")
+		}
+
+		// add item key to monitors map // add watch to the list
+		s.monitors[key] = m
+
+		klog.Info("secret informer started", " item key ", key)
+	}
+
+	// add the event handler
+	registration, err := m.itemMonitor.AddEventHandler(handler)
+	if err != nil {
+		return nil, err
+	}
+	m.numHandlers.Add(1)
+	klog.Info("secret handler added", " item key ", key)
+
+	return registration, nil
+}
+
+// RemoveSecretEventHandler removes a secret event handler and stops the informer if no handlers are left.
+// If the handler is not found or if there is an issue removing it, an error is returned.
+func (s *secretMonitor) RemoveSecretEventHandler(handlerRegistration SecretEventHandlerRegistration) error {
```

Comment:
take context

Reply:
Here also, do we need this for future usage? Unlike `AddSecretEventHandler()` where we need a context to stop the informer, I'm wondering what the context would do during removal.

Reply:
> Do we need this for future usage? Unlike `AddSecretEventHandler()` where we need a context to stop the informer, I'm wondering what the context would do during removal.

Cancellation and timeouts.  I won't stick on it.